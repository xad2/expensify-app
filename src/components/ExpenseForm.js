import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


const now = moment();

export default class ExpenseForm extends React.Component {
  
  constructor(props) {
    super(props); 
    
    this.state = { // might be getting from EditExpense or AddExpense
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? props.expense.amount.toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: '',
      editMode: props.expense ? true : false
    };
  }


  onDescrChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    // e.persist();
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if(!amount || amount.match(/^\d+(\.\d{0,2})?$/)){
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({createdAt}));
    }
  };

  onFocusChange = ({focused}) => {
    this.setState(() => ({calendarFocused: focused}));
  };

  onSubmit = (e) => {
    e.preventDefault();
    if(!this.state.description || !this.state.amount) {
      // set errpr state equal to 
      this.setState(() => ({error: 'Please provide description and amount'}));
    } else {
      this.setState(() => ({error: ''}));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount).toFixed(2),
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <span >{this.state.error}</span>}
        <form onSubmit={this.onSubmit}>
          <input 
            type="text"
            placeholder="Description"
            autoFocus 
            value={this.state.description}
            onChange={this.onDescrChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker 
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={ () => false }
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
            />
          <button type="submit">{this.state.editMode ? 'Edit Expense' : 'Add Expense'}</button>
        </form>
      </div>
    );
  };
}