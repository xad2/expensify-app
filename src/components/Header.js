import React from 'react';
import {Link, NavLink} from 'react-router-dom';
const Header = () => (
    <header>
      <h1>Expensify</h1>
      <ul>
        <li><NavLink to="/" activeClassName="is-active" exact>Home</NavLink></li>
        <li><NavLink to="/create" activeClassName="is-active">Create Expense</NavLink></li>
        <li><NavLink to="/help" activeClassName="is-active">Help</NavLink></li>
      </ul>
    </header>
  );

export default Header;