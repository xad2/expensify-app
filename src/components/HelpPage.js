import React from 'react';

const HelpPage = () => (
    <div>
      <p>HelpPageComp</p>
    </div>
  );

  export default HelpPage;