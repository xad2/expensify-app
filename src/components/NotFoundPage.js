import React from 'react';
import {BrowserRouter, Route, Switch, Link, NavLink} from 'react-router-dom';
const NotFoundPage = () => (
    <div>
      <p>404! - <Link to="/">Go home</Link></p> 
    </div>
  );

  export default NotFoundPage;