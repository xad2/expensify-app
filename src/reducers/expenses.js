// Expenses Reducer

const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      console.log('ADD', action.expense);
      return [
        ...state,
        action.expense
      ];
    case 'REMOVE_EXPENSE':
      console.log('REMOVE', action.expense);
      return state.filter( (expense) => 
        expense.id !== action.id 
      );
    case 'EDIT_EXPENSE':
      return state.map( (expense) => {
        if( expense.id === action.id ) {
          console.log('EDIT', action.updates);
          return {
            ...expense,
            ...action.updates // spread object
          };
        } else {
          return expense;
        };
      });
    default: return state;
  }
};

export default expensesReducer;