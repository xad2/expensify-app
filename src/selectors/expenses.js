import moment from 'moment';

// Get visible expenses (filter)
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense, index) => {
    const createdAtMoment = moment(expense.createdAt);
    
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;

    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;

    const textMatch = expense.description.toLowerCase().trim().includes(text.toLowerCase().trim());


    return startDateMatch && endDateMatch && textMatch;
  }).sort((a,b) => { // if return is -1 put a before b, else if a return 1 put b before a, else return same 
    if (sortBy === 'date'){ // if a < b return 1, else a > b return 1, else a = b return 0
      return a.createdAt < b.createdAt? 1 : -1; // from latest 
    }else if( sortBy === 'amount'){
      return a.amount < b.amount? 1 : -1; // from greatest to lowest
    }
  });
};

export default getVisibleExpenses;