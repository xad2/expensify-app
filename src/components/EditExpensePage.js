import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    // this.props.editExpense(this.props.match.params.id, expense);
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  onClickRemove = () => {
    this.props.removeExpense(this.props.expense);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
      <p>Edit the expense with id of {this.props.expense.id}</p>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button type="submit" onClick={ this.onClickRemove }>Remove</button>
      </div>
    );   
  };
}



const mapStateToProps = (state, props) => { // gives ExpenseForm component current state to this.props
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch) => ({
  removeExpense: (expense) => dispatch(removeExpense(expense)),
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
});
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);