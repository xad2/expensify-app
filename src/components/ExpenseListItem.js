import React from 'react';
// import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';
import {Link, NavLink} from 'react-router-dom';

export default (
  ({expense}) => {
  // (props) => {
    return (
      <div>
        <div><Link to={`/edit/${expense.id}`} >{expense.description}</Link></div>
        <div>{expense.amount}</div>
        <div>{expense.createdAt}</div>        
      </div>
    );
  }
);

 