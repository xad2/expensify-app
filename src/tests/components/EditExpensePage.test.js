import React from 'react';
import { shallow } from 'enzyme';
import {EditExpensePage} from '../../components/EditExpensePage';
import expenses from '../dummydata/expenses';

let editExpense, removeExpense, expense, history, wrapper;
beforeEach( () => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn()};
  expense = expenses[2];
  wrapper = shallow(<EditExpensePage expense={expense} editExpense={editExpense} removeExpense={removeExpense} history={history}/>);
});
test('should render editExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});
test('should handle editExpense correctly', () => {
  const editedExpense = {description: 'edited expense', id: '3'};
  wrapper.find('ExpenseForm').prop('onSubmit')(editedExpense);
  expect(editExpense).toHaveBeenCalledWith(editedExpense.id, editedExpense);
  expect(history.push).toHaveBeenCalledWith('/');
  expect(wrapper).toMatchSnapshot();
});
test('should handle removeExpense correctly', () => {
  wrapper.find('button').simulate('click');
  expect(removeExpense).toHaveBeenCalledWith(expense);
  expect(history.push).toHaveBeenCalledWith('/');
  expect(wrapper).toMatchSnapshot();
});