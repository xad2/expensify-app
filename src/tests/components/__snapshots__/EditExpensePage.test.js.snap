// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle editExpense correctly 1`] = `
<div>
  <p>
    Edit the expense with id of 
    3
  </p>
  <ExpenseForm
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
    type="submit"
  >
    Remove
  </button>
</div>
`;

exports[`should handle removeExpense correctly 1`] = `
<div>
  <p>
    Edit the expense with id of 
    3
  </p>
  <ExpenseForm
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
    type="submit"
  >
    Remove
  </button>
</div>
`;

exports[`should render editExpensePage correctly 1`] = `
<div>
  <p>
    Edit the expense with id of 
    3
  </p>
  <ExpenseForm
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
    type="submit"
  >
    Remove
  </button>
</div>
`;
